### ISSUE TITLE
Error in syntax highlighting function call (regression)

### ISSUE THREAD

**Issue by frangio** [neovim/neovim#20456]
### Neovim version (nvim -v)

NVIM v0.8.0-1210-gd367ed9b2

### Vim (not Nvim) behaves the same?

no, vim 9.0.354

### Operating system/version

Linux 5.19.12-arch1-1

### Terminal name/version

alacritty 0.10.1

### $TERM environment variable

alacritty

### Installation

Arch Linux Pacman

### How to reproduce the issue

Create a test.vim file with the following contents and open it.

```vim
lua <<EOF
print(a)
EOF

echo vim
```

### Expected behavior

Within `<<EOF ... EOF` syntax is highlighted as Lua, after `EOF` highlighting returns to Vim.

Neovim v0.7.2:

![image](https://user-images.githubusercontent.com/481465/193473919-c785df19-4238-494b-a222-86669d2df1bb.png)

### Actual behavior

The closing parenthesis is matched as `luaParenError`. The error affects the remainder of the file, in particular `EOF` doesn't mark the end of the Lua highlighting.

Neovim v0.8.0:

![image](https://user-images.githubusercontent.com/481465/193473937-2abdcb14-5669-4aae-b950-980ce6ca483d.png)

I've seen this affect [other languages](https://github.com/TovarishFin/vim-solidity) as well so it doesn't appear to be a problem with the specific syntax files.

**Comment by clason** [comment]
I cannot reproduce this. Neovim uses the exact same Lua syntax file as Vim, so this is highly unlikely to be an actual bug.

Can you reproduce this with Nvim 0.8 and `nvim --clean`?

**Comment by zeertzjq** [comment]
This should be reported to https://github.com/marcuscf/vim-lua

**Comment by frangio** [comment]
> Can you reproduce this with Nvim 0.8 release and `nvim --clean`?

Yes, my issue was with Nvim 0.8 release and `nvim -u NORC`.

I don't think this issue is with marcuscf/vim-lua... Using the latest syntax file from that repo I can reproduce the issue in Nvim 0.8 but not in 0.7.

**Comment by clason** [comment]
I can now reproduce this issue with Vim 9.0.0650, so this _is_ an upstream bug with the syntax files. Please report this at vim/vim (as this may be related to the injection in VimL and not the Lua syntax file).

**Comment by zeertzjq** [comment]
This is a bug in syntax highlighting.

**Comment by clason** [comment]
Ah. Good thing we have treesitter for these languages now, then ;)

**Comment by mg979** [comment]
It's lua syntax file that was 'updated'. The one shipped with 0.7.2 worked well.

>This is a bug in Vim's syntax engine, not syntax files

Can be but why ship a bugged syntax file with the stable release.

>Ah. Good thing we have treesitter for these languages now, then ;)

Not a good reason to break the old syntax files.

**Comment by wookayin** [comment]
What's our plan to fix this regression on the lua syntax file? For the 0.8.1 hotfix, Would it be fine to simply revert to the version at 91a2e7a57 (which works fine) if the upstream fix would take quite long? With some bisect, it can be found that the syntax was broken in 67df3347f (#20240).

I was thinking to reporting an issue to @marcuscf's vim-lua upstream, but I actually cannot find [a version in nvim-lua](https://github.com/marcuscf/vim-lua/commits/master/syntax/lua.vim) that matches 91a2e7a57 (last updated 2022-Mar-31).

**Comment by frangio** [comment]
Just to clarify for others because I was personally confused... The syntax file itself is not bugged, but there is a bug in the syntax engine (present in 0.8.0 but not in 0.7.2) that is triggered by this new syntax file and not the previous one.

**Comment by zeertzjq** [comment]
The bug in syntax engine is present in versions as early as v0.5.1.

**Comment by craigmac** [comment]
In case someone comes across this before it's fixed, to 'fix' (by just turning off that bad highlighting) this for now, add to your `init.vim`: 

```
" until this issue solved: https://github.com/neovim/neovim/issues/20456
augroup MyLuaHLBugFix
  autocmd!
  autocmd! ColorScheme,VimEnter * highlight! link luaParenError Normal | highlight! link luaError Normal
augroup END
```

**Comment by doronbehar** [comment]
> In case someone comes across this before it's fixed, to 'fix' (by just turning off that bad highlighting) this for now [..]

Your workaround did not work for me :(.

**Comment by cpmsmith** [comment]
> Ah. Good thing we have treesitter for these languages now, then ;)

To expand on this for the other noobs in the room like myself, this is referring to a (quasi-) workaround. If you have [nvim-treesitter](https://github.com/nvim-treesitter/nvim-treesitter) and the appropriate modules installed, you can use it for syntax highlighting instead of `vim-lua`. 

[Details are in nvim-treesitter's readme](https://github.com/nvim-treesitter/nvim-treesitter#modules), this is the TL;DR:
```lua
require('nvim-treesitter.configs').setup({
	highlight = {
		enable = true,
		disable = function(lang, bufnr)
			-- including vim for the sake of Lua blocks inside VimL
			if lang == "vim" or lang == "lua" then
				return false
			end
			return true
		end
	},
})
```

**Comment by clason** [comment]
No, this is completely handled by Neovim core. All you need is to put `vim.treesitter.start()` in a `ftplugin/lua.lua`

**Comment by craigmac** [comment]
@cpmsmith see https://github.com/neovim/neovim/pull/15391

**Comment by craigmac** [comment]
@chrisgrieser it just depends on what group you are linking it to, your solution is not going to work for me, for example. It entirely depends on your active colorscheme. Just look at `:hi` and link it to one that looks 'normal' (non bright red background).

**Comment by wookayin** [comment]
@clason suggested another great workaround (https://github.com/neovim/neovim/pull/21347#issuecomment-1343483802):

```
vim.g.ts_highlight_lua = true
```

Having the line somewhere in your `init.lua` will disable the vimscript syntax: syntax highlighting will be provided by treesitter (if the plugin exists).

Although some users may not prefer treesitter highlights, but it might be better than having a broken syntax. Or if you don't upgrade neovim that often (e.g. nightly), [manually overwriting the $VIMRUNTIME syntax file](https://github.com/neovim/neovim/issues/20456#issuecomment-1301427680) will be another good workaround.

**Comment by pencilcheck** [comment]
Is there a solution for those who still uses init.vim?

**Comment by clason** [comment]
You can use treesitter highlighting with `init.vim` just fine.

**Comment by pencilcheck** [comment]
> lua vim.treesitter.start()

Gotcha, I did this and it seems to fixed the highlighting for lua portion (but it changes the rest of the highlight for the .vim files though

```
--- inside init.vim
autocmd FileType vim lua vim.treesitter.start()
lua << EOF
vim.g.ts_highlight_lua = true
EOF
```

**Comment by clason** [comment]
you don't need the `vim` autocommand, or the `g:ts_highlight_lua` variable (which will be removed in 0.9)

**Comment by pappasam** [comment]
This appears to be resolved for me using the latest nightly + treesitter

**Comment by clason** [comment]
Yes, because the issue is strictly about the legacy syntax engine, which is disabled when you use treesitter (for that language). So the issue is still there; you're just not seeing it.

**Comment by Rmano** [comment]
Hi, still here. I tried to use the older lua syntax file but it seems it's not working anymore... 

![Image](https://github.com/user-attachments/assets/e40ae10e-4020-43fd-92f8-312cbfe10759)
